crates/vm/src/arch/interpreter_preflight.rs

```diff
@@ -1,31 +1,119 @@
-use openvm_stark_backend::p3_field::PrimeField32;
+use std::sync::Arc;
+
+use openvm_instructions::{instruction::Instruction, program::Program, LocalOpcode, SystemOpcode};
+use openvm_stark_backend::{
+    p3_field::{Field, PrimeField32},
+    p3_maybe_rayon::prelude::*,
+};
 
-use super::ExecutionError;
 use crate::{
     arch::{
-        execution_mode::PreflightCtx, instructions::*, Arena, PreflightExecutor, VmExecState,
-        VmStateMut,
+        execution_mode::PreflightCtx, interpreter::get_pc_index, Arena, ExecutionError, ExecutorId,
+        ExecutorInventory, PreflightExecutor, StaticProgramError, VmExecState, VmStateMut,
     },
-    system::{memory::online::TracingMemory, program::ProgramHandler},
+    system::memory::online::TracingMemory,
 };
 
+/// VM preflight executor (E3 executor) for use with trace generation.
+/// Note: This executor doesn't hold any VM state and can be used for multiple execution.
 pub struct PreflightInterpretedInstance<F, E> {
-    pub handler: ProgramHandler<F, E>,
-    executor_idx_to_air_idx: Vec<usize>,
+    // NOTE[jpw]: we use an Arc so that VmInstance can hold both VirtualMachine and
+    // PreflightInterpretedInstance. All we really need is to borrow `executors: &'a [E]`.
+    inventory: Arc<ExecutorInventory<E>>,
+
+    /// This is a map from (pc - pc_base) / pc_step -> [PcEntry].
+    /// We will map to `u32::MAX` if the program has no instruction at that pc.
+    // PERF[jpw/ayush]: We could map directly to the raw pointer(u64) for executor, but storing the
+    // u32 may be better for cache efficiency.
```

Comment:
unrelated, but i think before that we should come up with something like precompute for e3 so that we don't have to store the full `Instruction<F>` struct

Reply:
yes, I tried this in that other PR; it was just too painful with our `AdapterTraceStep` trait as-is :(